### PR TITLE
Store information about dimensions that are dropped in SlicedLowLevelWCS

### PR DESCRIPTION
**Pull request by Cadair**
~This is based on #9736 because otherwise they would conflict horrifically but they are actually orthogonal changes, it's just the last commit that needs to be looked at here.~

Now based on #10229 

The objective of this is to provide a structured way of inspecting dimensions that are dropped from a WCS when it is sliced. Take for example a spectral cube (space, space, wave), and you slice down to a two dimensional WCS (space, space) you loose access to the wavelength information for that slice. This PR implements a property `dropped_world_dimensions` onto `SlicedLowLevelWCS` which returns a dict containing the information hidden from the low level API (so you could construct a pixel_n_dim=0 Low Level WCS from it in theory).

### CODE REVIEW

File: astropy/wcs/wcsapi/wrappers/sliced_wcs.py
Comment (astrofrog) [inline astropy/wcs/wcsapi/wrappers/sliced_wcs.py:185]: Is there a reason for doing this here rather than in the initializer? This will have to be computed every time ``dropped_world_dimensions`` is called? Alternatively we could use @lazyproperty?

Comment (Cadair) [reply astropy/wcs/wcsapi/wrappers/sliced_wcs.py:185]: It seemed cleaner to do it in a property. Lazy property seems like a good idea. :+1:

**Comment by astrofrog** [comment]
I'll hold off from reviewing this until #9736 is merged and this is rebased

**Comment by mhvk** [comment]
This doesn't seem release critical, so moved the milestone (it does seem nice, though! but doing it as a step-by-step process through the development for 4.2 might be better anyway).

**Comment by Cadair** [comment]
> Looks nice! Just a small comment below. I'll re-review once this is rebased.

Well it's only like 6 months later but I rebased it!

**Review by astrofrog — CHANGES_REQUESTED** [review]
Looks good! But needs a changelog entry 😄

**Review by astrofrog — CHANGES_REQUESTED** [review]
Can you rebase so I can check the CI is all passing?

**Comment by Cadair** [comment]
@astrofrog build passed.

**Review by astrofrog — APPROVED** [review]
Looks great, thanks!

**Comment by pllim** [comment]
@Cadair , you broke pyinstaller job.

```
_________________ ERROR at setup of test_dropped_dimensions_4d _________________
file /home/travis/build/astropy/astropy/.pyinstaller/astropy_tests/wcs/wcsapi/wrappers/tests/test_sliced_wcs.py, line 849
  def test_dropped_dimensions_4d(cube_4d_fitswcs):
E       fixture 'cube_4d_fitswcs' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, header_time_1d, ignore_matplotlibrc, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, time_1d_wcs, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
/home/travis/build/astropy/astropy/.pyinstaller/astropy_tests/wcs/wcsapi/wrappers/tests/test_sliced_wcs.py:849
```

https://travis-ci.com/github/astropy/astropy/jobs/404316416

**Pull request merged**
